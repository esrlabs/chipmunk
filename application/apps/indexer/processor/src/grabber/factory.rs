use crate::{
    dlt_source::DltSource,
    grabber::{
        AsyncGrabTrait, ComputationResult, GrabError, GrabMetadata, Grabber, MetadataSource,
    },
    text_source::TextFileSource,
};
use serde::Serialize;
use std::path::Path;
use tokio_util::sync::CancellationToken;

#[derive(Debug, Serialize, Clone)]
pub enum SupportedFileType {
    Text,
    Dlt,
}

pub fn get_supported_file_type(path: &Path) -> Result<SupportedFileType, GrabError> {
    let extension = path.extension().map(|ext| ext.to_string_lossy());
    match extension {
        Some(ext) => match ext.to_lowercase().as_ref() {
            "dlt" => Ok(SupportedFileType::Dlt),
            "txt" | "text" => Ok(SupportedFileType::Text),
            _ => Ok(SupportedFileType::Text),
        },
        None => {
            // try to interpret as text
            if TextFileSource::contains_text(path)? {
                Ok(SupportedFileType::Text)
            } else {
                Err(GrabError::Unsupported(format!(
                    "Unsupported file type for {:?}",
                    path
                )))
            }
        }
    }
}

pub fn create_lazy_grabber(
    input_p: &Path,
    source_id: &str,
) -> Result<Box<dyn AsyncGrabTrait>, GrabError> {
    match get_supported_file_type(input_p)? {
        SupportedFileType::Text => {
            type GrabberType = Grabber<TextFileSource>;
            let source = TextFileSource::new(input_p, source_id);
            let grabber = GrabberType::lazy(source).map_err(|e| {
                let err_msg = format!("Could not create grabber: {}", e);
                warn!("{}", err_msg);
                GrabError::Config(err_msg)
            })?;
            Ok(Box::new(grabber))
        }
        SupportedFileType::Dlt => {
            type GrabberType = Grabber<DltSource>;
            let source = DltSource::new(input_p, source_id, true);
            let grabber = GrabberType::lazy(source)
                .map_err(|e| GrabError::Config(format!("Could not create grabber: {}", e)))?;
            Ok(Box::new(grabber))
        }
    }
}

pub fn create_metadata_for_source(
    file_path: &Path,
    source_id: String,
    cancellation_token: CancellationToken,
) -> Result<ComputationResult<GrabMetadata>, GrabError> {
    match get_supported_file_type(file_path)? {
        SupportedFileType::Dlt => {
            let source = DltSource::new(file_path, &source_id, true);
            source.from_file(Some(cancellation_token))
        }
        SupportedFileType::Text => {
            let source = TextFileSource::new(file_path, &source_id);
            source.from_file(Some(cancellation_token))
        }
    }
}
