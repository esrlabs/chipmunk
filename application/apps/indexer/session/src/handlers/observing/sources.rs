use crate::{
    events::{NativeError, NativeErrorKind},
    state::SessionStateAPI,
};
use indexer_base::progress::Severity;
use sources::{
    command,
    factory::{
        ParserType, ProcessTransportConfig, SerialTransportConfig, TCPTransportConfig, Transport,
        UDPTransportConfig,
    },
    pcap, raw, serial, socket, ByteSource,
};
use std::{fs::File, path::PathBuf};

pub async fn stream(transport: &Transport) -> Result<Box<dyn ByteSource>, NativeError> {
    match transport {
        Transport::UDP(settings) => udp(settings).await,
        Transport::TCP(settings) => tcp(settings).await,
        Transport::Serial(settings) => serial(settings).await,
        Transport::Process(settings) => cmd(settings).await,
    }
}

pub fn file(parser: &ParserType, filename: &PathBuf) -> Result<Box<dyn ByteSource>, NativeError> {
    match parser {
        ParserType::SomeIP(_) => Err(NativeError {
            severity: Severity::ERROR,
            kind: NativeErrorKind::FileNotFound,
            message: Some(String::from("SomeIP parser not yet supported")),
        }),
        ParserType::Pcap(_) => pcap(filename),
        ParserType::Dlt(_) => binary(filename),
        ParserType::Text => binary(filename),
    }
}

fn binary(filename: &PathBuf) -> Result<Box<dyn ByteSource>, NativeError> {
    Ok(Box::new(raw::binary::BinaryByteSource::new(
        File::open(filename).map_err(|e| NativeError {
            severity: Severity::ERROR,
            kind: NativeErrorKind::Io,
            message: Some(format!(
                "Fail open file {}: {}",
                filename.to_string_lossy(),
                e
            )),
        })?,
    )))
}

fn pcap(filename: &PathBuf) -> Result<Box<dyn ByteSource>, NativeError> {
    Ok(Box::new(
        pcap::file::PcapngByteSource::new(File::open(filename).map_err(|e| NativeError {
            severity: Severity::ERROR,
            kind: NativeErrorKind::Io,
            message: Some(format!(
                "Fail open file {}: {}",
                filename.to_string_lossy(),
                e
            )),
        })?)
        .map_err(|e| NativeError {
            severity: Severity::ERROR,
            kind: NativeErrorKind::ComputationFailed,
            message: Some(format!(
                "Fail create source for {}: {}",
                filename.to_string_lossy(),
                e
            )),
        })?,
    ))
}

async fn udp(config: &UDPTransportConfig) -> Result<Box<dyn ByteSource>, NativeError> {
    Ok(Box::new(
        socket::udp::UdpSource::new(&config.bind_addr, config.multicast.clone())
            .await
            .map_err(|e| NativeError {
                severity: Severity::ERROR,
                kind: NativeErrorKind::Interrupted,
                message: Some(format!("Fail to create socket due error: {:?}", e)),
            })?,
    ))
}

async fn tcp(config: &TCPTransportConfig) -> Result<Box<dyn ByteSource>, NativeError> {
    Ok(Box::new(
        socket::tcp::TcpSource::new(config.bind_addr.clone())
            .await
            .map_err(|e| NativeError {
                severity: Severity::ERROR,
                kind: NativeErrorKind::Interrupted,
                message: Some(format!("Fail to create socket due error: {:?}", e)),
            })?,
    ))
}

async fn cmd(config: &ProcessTransportConfig) -> Result<Box<dyn ByteSource>, NativeError> {
    Ok(Box::new(
        command::process::ProcessSource::new(
            config.command.clone(),
            config.args.clone(),
            config.envs.clone(),
        )
        .await
        .map_err(|e| NativeError {
            severity: Severity::ERROR,
            kind: NativeErrorKind::Interrupted,
            message: Some(format!("Fail to create process source due error: {:?}", e)),
        })?,
    ))
}

async fn serial(config: &SerialTransportConfig) -> Result<Box<dyn ByteSource>, NativeError> {
    Ok(Box::new(
        serial::serialport::SerialSource::new(config).map_err(|e| NativeError {
            severity: Severity::ERROR,
            kind: NativeErrorKind::Interrupted,
            message: Some(format!(
                "Fail to create serial connection due error: {:?}",
                e
            )),
        })?,
    ))
}

pub async fn get_source_id(state: &SessionStateAPI, uuid: &str) -> Result<u8, NativeError> {
    state.add_source(uuid).await
}
