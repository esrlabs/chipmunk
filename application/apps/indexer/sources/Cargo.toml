[package]
name = "sources"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
async-stream = "0.3"
async-trait = "0.1"
buf_redux = "0.8"
crossbeam-channel = "0.5"
dlt-core = "0.11"
etherparse = "0.10"
futures = "0.3"
indexer_base = { path = "../indexer_base" }
log = "0.4"
pcap-parser = "0.13"
pin-project-lite = "0.2"
thiserror = "1.0"
tokio = { version = "1.5", features = ["full"] }
tokio-stream = "0.1.5"
tokio-util = "0.6.9"

[dev-dependencies]
lazy_static = "1.4"
env_logger = "0.9"