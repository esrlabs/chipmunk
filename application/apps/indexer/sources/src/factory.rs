use indexer_base::config::MulticastInfo;
use parsers::dlt;
use serde::{Deserialize, Serialize};
use std::{collections::HashMap, path::PathBuf};
use uuid::Uuid;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub enum ParserType {
    Dlt(DltParserSettings),
    SomeIP(SomeIPParserSettings),
    Text,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct DltParserSettings {
    pub filter_config: Option<dlt::DltFilterConfig>,
    pub fibex_file_paths: Option<Vec<String>>,
    pub with_storage_header: bool,
    #[serde(skip)]
    pub fibex_metadata: Option<dlt::FibexMetadata>,
}

impl Default for DltParserSettings {
    fn default() -> Self {
        Self {
            filter_config: None,
            fibex_file_paths: None,
            with_storage_header: true,
            fibex_metadata: None,
        }
    }
}

impl DltParserSettings {
    pub fn new_including_storage_headers(
        filter_config: Option<dlt::DltFilterConfig>,
        fibex_file_paths: Option<Vec<String>>,
    ) -> Self {
        Self {
            filter_config,
            fibex_file_paths,
            with_storage_header: true,
            fibex_metadata: None,
        }
    }

    pub fn load_fibex_metadata(&mut self) {
        if self.fibex_metadata.is_some() {
            return;
        }
        self.fibex_metadata = if let Some(paths) = self.fibex_file_paths.as_ref() {
            dlt::gather_fibex_data(dlt::FibexConfig {
                fibex_file_paths: paths.clone(),
            })
        } else {
            None
        };
    }
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct PcapParserSettings {
    pub dlt: DltParserSettings,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct SomeIPParserSettings {}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub enum Transport {
    Process(ProcessTransportConfig),
    TCP(TCPTransportConfig),
    UDP(UDPTransportConfig),
    Serial(SerialTransportConfig),
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ProcessTransportConfig {
    pub cwd: PathBuf,
    pub command: String,
    pub envs: HashMap<String, String>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct SerialTransportConfig {
    pub path: String,
    pub baud_rate: u32,
    pub data_bits: u8,
    pub flow_control: u8,
    pub parity: u8,
    pub stop_bits: u8,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct TCPTransportConfig {
    pub bind_addr: String,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct UDPTransportConfig {
    pub bind_addr: String,
    pub multicast: Vec<MulticastInfo>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub enum ObserveFileType {
    Pcap,
    Text,
    Binary,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub enum ObserveOrigin {
    File(String, ObserveFileType, PathBuf),
    Concat(Vec<(String, ObserveFileType, PathBuf)>),
    Stream(String, Transport),
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ObserveOptions {
    pub origin: ObserveOrigin,
    pub parser: ParserType,
}

impl ObserveOptions {
    pub fn file(filename: PathBuf, file_origin: ObserveFileType, parser: ParserType) -> Self {
        ObserveOptions {
            origin: ObserveOrigin::File(Uuid::new_v4().to_string(), file_origin, filename),
            parser,
        }
    }
}
