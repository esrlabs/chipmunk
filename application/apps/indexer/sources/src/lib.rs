use std::io::Read;

use components::Components;
use definitions::{ReloadInfo, SourceError, SourceFilter};

// Rust can't currently distinguish between dev and none-dev dependencies at the moment. There is
// an open issue for this case: "https://github.com/rust-lang/rust/issues/129637"
#[macro_use]
extern crate lazy_static;

#[macro_use]
extern crate log;

pub mod binary;
pub mod command;
pub mod prelude;
pub mod serial;
pub mod socket;

pub enum Source {
    Raw(binary::raw::BinaryByteSourceFromFile),
    Pcap(binary::pcap::legacy::PcapLegacyByteSourceFromFile),
    PcapNg(binary::pcap::ng::PcapngByteSourceFromFile),
    Tcp(socket::tcp::TcpSource),
    Udp(socket::udp::UdpSource),
    Serial(serial::serialport::SerialSource),
    Process(command::process::ProcessSource),
}

impl definitions::ByteSource for Source {
    fn consume(&mut self, offset: usize) {
        match self {
            Self::Raw(inner) => inner.consume(offset),
            Self::Pcap(inner) => inner.consume(offset),
            Self::PcapNg(inner) => inner.consume(offset),
            Self::Tcp(inner) => inner.consume(offset),
            Self::Udp(inner) => inner.consume(offset),
            Self::Serial(inner) => inner.consume(offset),
            Self::Process(inner) => inner.consume(offset),
        }
    }

    fn current_slice(&self) -> &[u8] {
        match self {
            Self::Raw(inner) => inner.current_slice(),
            Self::Pcap(inner) => inner.current_slice(),
            Self::PcapNg(inner) => inner.current_slice(),
            Self::Tcp(inner) => inner.current_slice(),
            Self::Udp(inner) => inner.current_slice(),
            Self::Serial(inner) => inner.current_slice(),
            Self::Process(inner) => inner.current_slice(),
        }
    }

    fn len(&self) -> usize {
        match self {
            Self::Raw(inner) => inner.len(),
            Self::Pcap(inner) => inner.len(),
            Self::PcapNg(inner) => inner.len(),
            Self::Tcp(inner) => inner.len(),
            Self::Udp(inner) => inner.len(),
            Self::Serial(inner) => inner.len(),
            Self::Process(inner) => inner.len(),
        }
    }

    async fn load(
        &mut self,
        filter: Option<&SourceFilter>,
    ) -> Result<Option<ReloadInfo>, SourceError> {
        match self {
            Self::Raw(inner) => inner.load(filter).await,
            Self::Pcap(inner) => inner.load(filter).await,
            Self::PcapNg(inner) => inner.load(filter).await,
            Self::Tcp(inner) => inner.load(filter).await,
            Self::Udp(inner) => inner.load(filter).await,
            Self::Serial(inner) => inner.load(filter).await,
            Self::Process(inner) => inner.load(filter).await,
        }
    }

    async fn cancel(&mut self) -> Result<(), SourceError> {
        match self {
            Self::Raw(inner) => inner.cancel().await,
            Self::Pcap(inner) => inner.cancel().await,
            Self::PcapNg(inner) => inner.cancel().await,
            Self::Tcp(inner) => inner.cancel().await,
            Self::Udp(inner) => inner.cancel().await,
            Self::Serial(inner) => inner.cancel().await,
            Self::Process(inner) => inner.cancel().await,
        }
    }

    async fn income(
        &mut self,
        msg: stypes::SdeRequest,
    ) -> Result<stypes::SdeResponse, SourceError> {
        match self {
            Self::Raw(inner) => inner.income(msg).await,
            Self::Pcap(inner) => inner.income(msg).await,
            Self::PcapNg(inner) => inner.income(msg).await,
            Self::Tcp(inner) => inner.income(msg).await,
            Self::Udp(inner) => inner.income(msg).await,
            Self::Serial(inner) => inner.income(msg).await,
            Self::Process(inner) => inner.income(msg).await,
        }
    }
}

// TODO: this registration function will fail if some of source would not be registred. That's wrong.
// If some of sources are failed, another sources should still be registred as well
pub fn registration<P>(components: &mut Components<Source, P>) -> Result<(), stypes::NativeError> {
    components.add_source(binary::raw::Descriptor::default())?;
    components.add_source(binary::pcap::legacy::Descriptor::default())?;
    components.add_source(binary::pcap::ng::Descriptor::default())?;
    components.add_source(socket::tcp::Descriptor::default())?;
    components.add_source(socket::udp::Descriptor::default())?;
    components.add_source(serial::options::Descriptor::default())?;
    components.add_source(command::process::Descriptor::default())?;
    Ok(())
}
