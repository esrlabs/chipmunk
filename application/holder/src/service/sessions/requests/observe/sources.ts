import { CancelablePromise } from 'platform/env/promise';
import { sessions } from '@service/sessions';
import { Logger } from 'platform/log';

import * as Requests from 'platform/ipc/request';

export const handler = Requests.InjectLogger<
    Requests.Observe.SourcesDefinitionsList.Request,
    CancelablePromise<Requests.Observe.SourcesDefinitionsList.Response>
>(
    (
        log: Logger,
        request: Requests.Observe.SourcesDefinitionsList.Request,
    ): CancelablePromise<Requests.Observe.SourcesDefinitionsList.Response> => {
        return new CancelablePromise((resolve, reject) => {
            const stored = sessions.get(request.session);
            if (stored === undefined) {
                return reject(new Error(`Session doesn't exist`));
            }
            stored.session
                .getStream()
                .getSourcesDefinitions()
                .then((sources) => {
                    resolve(
                        new Requests.Observe.SourcesDefinitionsList.Response({
                            session: stored.session.getUUID(),
                            sources,
                        }),
                    );
                })
                .catch(reject);
        });
    },
);
