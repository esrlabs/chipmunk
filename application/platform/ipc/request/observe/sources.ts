import { Define, Interface, SignatureRequirement } from '../declarations';
import { ISourceLink } from '../../../types/observe/types/index';

import * as validator from '../../../env/obj';

@Define({ name: 'SourcesListRequest' })
export class Request extends SignatureRequirement {
    public session: string;

    constructor(input: { session: string }) {
        super();
        validator.isObject(input);
        this.session = validator.getAsNotEmptyString(input, 'session');
    }
}

export interface Request extends Interface {}

@Define({ name: 'SourcesListResponse' })
export class Response extends SignatureRequirement {
    public session: string;
    public sources: ISourceLink[];

    constructor(input: { session: string; sources: ISourceLink[] }) {
        super();
        validator.isObject(input);
        this.session = validator.getAsNotEmptyString(input, 'session');
        this.sources = validator.getAsArray<ISourceLink>(input, 'sources');
    }
}

export interface Response extends Interface {}
