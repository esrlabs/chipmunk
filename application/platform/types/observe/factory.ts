import * as $ from './index';

export { FileType } from './types/file/index';

class Factory<T> {
    protected changes: { parser: boolean; origin: boolean } = { parser: false, origin: false };

    protected updated(): {
        parser(): void;
        origin(): void;
    } {
        const update = () => {
            if (this.changes.origin === this.changes.parser) {
                // No needs to update if and parser and origin were changed or weren't changed both
                return;
            }
            if (this.changes.origin) {
                const parsers = this.observe.origin.getSupportedParsers();
                if (parsers.length === 0) {
                    throw new Error(
                        `No supported parser for ${this.observe.origin.instance.alias()}`,
                    );
                }
                this.protocol(parsers[0].alias());
            } else if (this.changes.parser) {
                // TODO: check source also
                // switch (this.observe.origin.instance.alias()) {
                //     case $.Origin.Context.File:
                //         const _filetypes = this.observe.parser.getSupportedFileType();
                //         break;
                //     case $.Origin.Context.Concat:
                //         const _filetypes = this.observe.parser.getSupportedFileType();
                //         break;
                //     case $.Origin.Context.Stream:
                //         const _streams = this.observe.parser
                //             .getSupportedStream()
                //             .map((r) => r.alias());
                //         break;
                // }
            }
        };
        return {
            parser: (): void => {
                this.changes.parser = true;
                update();
            },
            origin: (): void => {
                this.changes.origin = true;
                update();
            },
        };
    }

    protected observe: $.Observe = $.Observe.new();

    public protocol(protocol: $.Parser.Protocol): T {
        this.observe.parser.change($.Parser.getByAlias(protocol));
        this.updated().parser();
        return this as unknown as T;
    }

    public asDlt(configuration?: $.Parser.Dlt.IConfiguration): T {
        this.observe.parser.change(
            new $.Parser.Dlt.Configuration(
                configuration === undefined ? $.Parser.Dlt.Configuration.initial() : configuration,
                undefined,
            ),
        );
        this.updated().parser();
        return this as unknown as T;
    }

    public asSomeip(configuration?: $.Parser.SomeIp.IConfiguration): T {
        this.observe.parser.change(
            new $.Parser.SomeIp.Configuration(
                configuration === undefined
                    ? $.Parser.SomeIp.Configuration.initial()
                    : configuration,
                undefined,
            ),
        );
        this.updated().parser();
        return this as unknown as T;
    }

    public asText(): T {
        this.observe.parser.change(new $.Parser.Text.Configuration(null, undefined));
        this.updated().parser();
        return this as unknown as T;
    }

    public get(): $.Observe {
        const parsers = this.observe.origin.getSupportedParsers().map((ref) => ref.alias());
        const selected = this.observe.parser.alias();
        if (!parsers.includes(selected)) {
            throw new Error(
                `Origin "${this.observe.origin.getNatureAlias()}" doesn't support parser: ${selected}; available parsers: ${parsers.join(
                    ', ',
                )}.`,
            );
        }
        return this.observe;
    }
}

export class File extends Factory<File> {
    static FileType = $.Types.File.FileType;

    constructor() {
        super();
        this.observe.origin.change(
            new $.Origin.File.Configuration($.Origin.File.Configuration.initial(), undefined),
        );
    }

    public source(id: string): File {
        if (!(this.observe.origin.instance instanceof $.Origin.File.Configuration)) {
            throw new Error(`Given observe object doesn't have File origin`);
        }
        this.observe.origin.instance.set().source(id);
        this.updated().origin();
        return this;
    }

    public file(filename: string): File {
        if (!(this.observe.origin.instance instanceof $.Origin.File.Configuration)) {
            throw new Error(`Given observe object doesn't have File origin`);
        }
        this.observe.origin.instance.set().filename(filename);
        this.updated().origin();
        return this;
    }

    public type(type: $.Types.File.FileType): File {
        if (!(this.observe.origin.instance instanceof $.Origin.File.Configuration)) {
            throw new Error(`Given observe object doesn't have File origin`);
        }
        this.observe.origin.instance.set().type(type);
        this.updated().origin();
        return this;
    }
}

export class Concat extends Factory<Concat> {
    static FileType = $.Types.File.FileType;

    constructor() {
        super();
        this.observe.origin.change(
            new $.Origin.Concat.Configuration($.Origin.Concat.Configuration.initial(), undefined),
        );
    }

    public files(files: string[]): Concat {
        if (!(this.observe.origin.instance instanceof $.Origin.Concat.Configuration)) {
            throw new Error(`Given observe object doesn't have Concat origin`);
        }
        this.observe.origin.instance.set().files(files);
        this.updated().origin();
        return this;
    }

    public type(type: $.Types.File.FileType): Concat {
        if (!(this.observe.origin.instance instanceof $.Origin.Concat.Configuration)) {
            throw new Error(`Given observe object doesn't have Concat origin`);
        }
        this.observe.origin.instance.set().defaults(type);
        this.updated().origin();
        return this;
    }

    public push(filename: string, type: $.Types.File.FileType): Concat {
        if (!(this.observe.origin.instance instanceof $.Origin.Concat.Configuration)) {
            throw new Error(`Given observe object doesn't have Concat origin`);
        }
        this.observe.origin.instance.set().push(filename, type);
        this.updated().origin();
        return this;
    }

    public remove(filename: string): Concat {
        if (!(this.observe.origin.instance instanceof $.Origin.Concat.Configuration)) {
            throw new Error(`Given observe object doesn't have Concat origin`);
        }
        this.observe.origin.instance.set().remove(filename);
        this.updated().origin();
        return this;
    }
}

export class Stream extends Factory<Stream> {
    constructor() {
        super();
        this.observe.origin.change(
            new $.Origin.Stream.Configuration($.Origin.Stream.Configuration.initial(), undefined),
        );
    }

    public process(configuration?: $.Origin.Stream.Stream.Process.IConfiguration): Stream {
        if (!(this.observe.origin.instance instanceof $.Origin.Stream.Configuration)) {
            throw new Error(`Given observe object doesn't have Stream origin`);
        }
        this.observe.origin.instance.change(
            new $.Origin.Stream.Stream.Process.Configuration(
                configuration !== undefined
                    ? configuration
                    : $.Origin.Stream.Stream.Process.Configuration.initial(),
                undefined,
            ),
        );
        this.updated().origin();
        return this;
    }

    public serial(configuration?: $.Origin.Stream.Stream.Serial.IConfiguration): Stream {
        if (!(this.observe.origin.instance instanceof $.Origin.Stream.Configuration)) {
            throw new Error(`Given observe object doesn't have Stream origin`);
        }
        this.observe.origin.instance.change(
            new $.Origin.Stream.Stream.Serial.Configuration(
                configuration !== undefined
                    ? configuration
                    : $.Origin.Stream.Stream.Serial.Configuration.initial(),
                undefined,
            ),
        );
        this.updated().origin();
        return this;
    }

    public tcp(configuration?: $.Origin.Stream.Stream.TCP.IConfiguration): Stream {
        if (!(this.observe.origin.instance instanceof $.Origin.Stream.Configuration)) {
            throw new Error(`Given observe object doesn't have Stream origin`);
        }
        this.observe.origin.instance.change(
            new $.Origin.Stream.Stream.TCP.Configuration(
                configuration !== undefined
                    ? configuration
                    : $.Origin.Stream.Stream.TCP.Configuration.initial(),
                undefined,
            ),
        );
        this.updated().origin();
        return this;
    }

    public udp(configuration?: $.Origin.Stream.Stream.UDP.IConfiguration): Stream {
        if (!(this.observe.origin.instance instanceof $.Origin.Stream.Configuration)) {
            throw new Error(`Given observe object doesn't have Stream origin`);
        }
        this.observe.origin.instance.change(
            new $.Origin.Stream.Stream.UDP.Configuration(
                configuration !== undefined
                    ? configuration
                    : $.Origin.Stream.Stream.UDP.Configuration.initial(),
                undefined,
            ),
        );
        this.updated().origin();
        return this;
    }
}

export function map(observe: $.Observe): {
    file(): boolean;
    concat(): boolean;
    process(): boolean;
    serial(): boolean;
    udp(): boolean;
    tcp(): boolean;
} {
    return {
        file: (): boolean => {
            return (
                observe.origin.as<$.Origin.File.Configuration>($.Origin.File.Configuration) !==
                undefined
            );
        },
        concat: (): boolean => {
            return (
                observe.origin.as<$.Origin.Concat.Configuration>($.Origin.Concat.Configuration) !==
                undefined
            );
        },
        process: (): boolean => {
            const stream = observe.origin.as<$.Origin.Stream.Configuration>(
                $.Origin.Stream.Configuration,
            );
            if (stream === undefined) {
                return false;
            }
            return (
                stream.as<$.Origin.Stream.Stream.Process.Configuration>(
                    $.Origin.Stream.Stream.Process.Configuration,
                ) !== undefined
            );
        },
        serial: (): boolean => {
            const stream = observe.origin.as<$.Origin.Stream.Configuration>(
                $.Origin.Stream.Configuration,
            );
            if (stream === undefined) {
                return false;
            }
            return (
                stream.as<$.Origin.Stream.Stream.Serial.Configuration>(
                    $.Origin.Stream.Stream.Serial.Configuration,
                ) !== undefined
            );
        },
        udp: (): boolean => {
            const stream = observe.origin.as<$.Origin.Stream.Configuration>(
                $.Origin.Stream.Configuration,
            );
            if (stream === undefined) {
                return false;
            }
            return (
                stream.as<$.Origin.Stream.Stream.UDP.Configuration>(
                    $.Origin.Stream.Stream.UDP.Configuration,
                ) !== undefined
            );
        },
        tcp: (): boolean => {
            const stream = observe.origin.as<$.Origin.Stream.Configuration>(
                $.Origin.Stream.Configuration,
            );
            if (stream === undefined) {
                return false;
            }
            return (
                stream.as<$.Origin.Stream.Stream.TCP.Configuration>(
                    $.Origin.Stream.Stream.TCP.Configuration,
                ) !== undefined
            );
        },
    };
}
