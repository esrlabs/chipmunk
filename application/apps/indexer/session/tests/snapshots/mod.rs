// TODO AAZ: Remove after prototyping
#![allow(unused)]

mod utls;

use pretty_assertions::{assert_eq, assert_ne};
use sources::factory::{DltParserSettings, FileFormat, ParserType};
use utls::*;

#[tokio::test]
async fn export_dlt() {
    let input = "../../../developing/resources/attachments.dlt";
    let session_file = run_export(
        input,
        FileFormat::Binary,
        ParserType::Dlt(DltParserSettings::default()),
    )
    .await;

    panic!("Session file: {}", session_file.display());
}
