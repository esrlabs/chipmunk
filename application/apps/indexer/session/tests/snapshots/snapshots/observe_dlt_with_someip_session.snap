---
source: session/tests/snapshots/mod.rs
description: Snapshot for DLT file with SomeIP network trace.
info:
  filter_config: ~
  fibex_file_paths:
    - "../../../developing/resources/someip.xml"
  with_storage_header: true
  tz: ~
snapshot_kind: text
---
session_file:
  - "2024-02-20T13:17:26.713537000Z\u0004ECU1\u00041\u0004571\u0004204\u000428138506\u0004ECU1\u0004APP1\u0004C1\u0004IPC\u0004SOME/IP RPC SERV:123 METH:32773 LENG:16 CLID:0 SEID:58252 IVER:1 MSTP:2 RETC:0 TestService::timeEvent {\u0006\ttimestamp (INT64) : 1683656786973,\u0006}"
  - "2024-02-20T13:17:26.713537000Z\u0004ECU1\u00041\u0004571\u0004205\u000428138506\u0004ECU1\u0004APP1\u0004C1\u0004IPC\u0004SOME/IP RPC SERV:124 METH:32773 LENG:16 CLID:0 SEID:58252 IVER:1 MSTP:2 RETC:0 UnknownService [00, 00, 01, 88, 01, C3, C4, 1D]"
  - "2024-02-20T13:17:26.713537000Z\u0004ECU1\u00041\u0004571\u0004206\u000428138506\u0004ECU1\u0004APP1\u0004C1\u0004IPC\u0004SOME/IP RPC SERV:123 METH:32773 LENG:16 CLID:0 SEID:58252 IVER:3 MSTP:2 RETC:0 TestService<1?>::timeEvent {\u0006\ttimestamp (INT64) : 1683656786973,\u0006}"
  - "2024-02-20T13:17:26.713537000Z\u0004ECU1\u00041\u0004571\u0004207\u000428138506\u0004ECU1\u0004APP1\u0004C1\u0004IPC\u0004SOME/IP RPC SERV:123 METH:32774 LENG:16 CLID:0 SEID:58252 IVER:1 MSTP:2 RETC:0 TestService::UnknownMethod [00, 00, 01, 88, 01, C3, C4, 1D]"
  - "2024-02-20T13:17:26.713537000Z\u0004ECU1\u00041\u0004571\u0004208\u000428138506\u0004ECU1\u0004APP1\u0004C1\u0004IPC\u0004SOME/IP RPC SERV:123 METH:32773 LENG:15 CLID:0 SEID:58252 IVER:1 MSTP:2 RETC:0 TestService::timeEvent 'SOME/IP Error: Parser exhausted at offset 0 for Object size 8' [00, 00, 01, 88, 01, C3, C4]"
  - "2024-02-20T13:17:26.713537000Z\u0004ECU1\u00041\u0004571\u0004209\u000428138506\u0004ECU1\u0004APP1\u0004C1\u0004IPC\u0004SOME/IP 'Parse error: Not enough data: min: 25, actual: 24' [00, 7B, 80, 05, 00, 00, 00, 11, 00, 00, E3, 8C, 01, 01, 02, 00, 00, 00, 01, 88, 01, C3, C4, 1D]"
attachments: []
